MigrateServicePreferenceFromLegacy/handler.ts

```diff
@@ -0,0 +1,136 @@
+}
+
+export const createServicePreference = (
+  serviceId: ServiceId,
+  blockedChannels: ReadonlyArray<BlockedInboxOrChannelEnum>,
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+): NewServicePreference => ({
+  fiscalCode,
+  id: makeServicesPreferencesDocumentId(fiscalCode, serviceId, version),
+  isEmailEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.EMAIL
+  ),
+  isInboxEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.INBOX
+  ),
+  isWebhookEnabled: !blockedChannels.some(
+    channel => channel === BlockedInboxOrChannelEnum.WEBHOOK
+  ),
+  kind: "INewServicePreference",
+  serviceId,
+  settingsVersion: version
+});
+
+export const blockedsToServicesPreferences = (
+  blocked: {
+    [x: string]: readonly BlockedInboxOrChannelEnum[];
+  },
+  fiscalCode: FiscalCode,
+  version: NonNegativeInteger
+) =>
+  o
+    .fromNullable(blocked)
+    .map(b =>
+      Object.entries(b)
+        .filter(i => ServiceId.is(i[0]))
+        .map(
+          i =>
+            createServicePreference(
+              i[0] as ServiceId,
+              i[1],
+              fiscalCode,
+              version
+            ) // cast required: ts do not identify filter as a guard
+        )
+    )
+    .getOrElse([]);
+
+export const MigrateServicePreferenceFromLegacy = (
+  servicePreferenceModel: ServicesPreferencesModel
+) => async (context: Context, input: unknown) =>
+  te
+    .fromEither(MigrateServicesPreferencesQueueMessage.decode(input))
+    .mapLeft(
+      ResponseErrorFromValidationErrors(MigrateServicesPreferencesQueueMessage)
+    )
+    .mapLeft(err => new Error(err.detail))
+    .filterOrElse(
+      migrateInput =>
+        NonNegativeInteger.is(
+          migrateInput.newProfile.servicePreferencesSettings.version
+        ),
+      new Error("Can not migrate to negative services preferences version.")
+    )
+    .map(migrateInput =>
+      blockedsToServicesPreferences(
+        migrateInput.oldProfile.blockedInboxOrChannels,
+        migrateInput.newProfile.fiscalCode,
+        /* tslint:disable-next-line no-useless-cast */
+        migrateInput.newProfile.servicePreferencesSettings
+          .version as NonNegativeInteger // cast required: ts do not identify filterOrElse as a guard
+      )
+    )
+    .map(preferences =>
+      preferences.map(preference =>
+        servicePreferenceModel
+          .create(preference)
+          .foldTaskEither<Error, boolean>(
+            cosmosError =>
+              isCosmosError(cosmosError) &&
+              cosmosError.error.code === CONFLICT_CODE
+                ? te.taskEither.of(false)
+                : te.fromLeft(
+                    new Error(
+                      `Can not create the service profile: ${JSON.stringify(
+                        cosmosError
+                      )}`
+                    )
+                  ),
+            _ => te.taskEither.of(true)
+          )
```

Comment:
Not sure if it's useful, maybe we can record the tuple serviceid-result 
(This code will break the tests)
```suggestion
          .foldTaskEither<Error, readonly [string, boolean]>(
            cosmosError =>
              isCosmosError(cosmosError) &&
              cosmosError.error.code === CONFLICT_CODE
                ? te.taskEither.of([preference.serviceId, false])
                : te.fromLeft(
                    new Error(
                      `Can not create the service profile: ${JSON.stringify(
                        cosmosError
                      )}`
                    )
                  ),
            _ => te.taskEither.of([preference.serviceId, true])
          )
```

Reply:
As steted above, any error will trigger the entire migration, so a single error is enough.